### PR TITLE
certificates: add K8s 1.8 impl

### PR DESCRIPTION
**Pull request by kopiczko**
Towards https://github.com/giantswarm/giantswarm/issues/1384

We need to get rid of kubernetes 1.7 dependencies completely in azure-operator. we also need to fetch cluster certs there. This is an attempt to achieve that.

This is heavily inspired from current certificatetpr package which this package deprecates.

### CODE REVIEW

File: spec.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package certificates
+
+type Interface interface {
+	SearchCluster(clusterID string) (Cluster, error)
+	SearchMonitoring(clusterID string) (Monitoring, error)
+}
```
Comment (kopiczko) [inline spec.go]: This is the new interface. Very simple & type-safe.

**Comment by kopiczko** [comment]
I don't find the diff useful (below I made a PR with the diff), but I'm happy
to exaplain what's going on. We can also meet.
Here's the diff: https://github.com/giantswarm/certificatetpr/pull/41/files

If you ignore vendor files the change there isn't much code.

* [error.go](https://github.com/giantswarm/certificates/blob/fa53ddad1cfcdc8d1e487700c42325e176bd1bc3/error.go)

Rather standard file.

* [spec.go](https://github.com/giantswarm/certificates/blob/fa53ddad1cfcdc8d1e487700c42325e176bd1bc3/spec.go)

Simple interface.

* [types.go](https://github.com/giantswarm/certificates/blob/fa53ddad1cfcdc8d1e487700c42325e176bd1bc3/types.go)

Structs and constants. No methods, no functions.

* [searcher.go](https://github.com/giantswarm/certificates/blob/fa53ddad1cfcdc8d1e487700c42325e176bd1bc3/searcher.go)

    - First 60 lines are boilerplate: https://github.com/giantswarm/certificates/blob/fa53ddad1cfcdc8d1e487700c42325e176bd1bc3/searcher.go#L1-L61
    - `search` body remained mostly unchanged. `case watch.Added` body was replaced with `fillTLSFromSecret` call: https://github.com/giantswarm/certificates/blob/fa53ddad1cfcdc8d1e487700c42325e176bd1bc3/searcher.go#L101-L140
    - This is new: https://github.com/giantswarm/certificates/blob/fa53ddad1cfcdc8d1e487700c42325e176bd1bc3/searcher.go#L63-L99
    - Reading the secred and filling TLS struct is new: https://github.com/giantswarm/certificates/blob/fa53ddad1cfcdc8d1e487700c42325e176bd1bc3/searcher.go#L142-L168


* [searcher_test.go](https://github.com/giantswarm/certificates/blob/fa53ddad1cfcdc8d1e487700c42325e176bd1bc3/searcher_test.go)

Tests for `fillTLSFromSecret`.

**Comment by xh3b4sd** [comment]
I just realized this is a new project. So we have `cert-operator`, and the `CertConfig` type for our generated clientset. Why is this called `certificates`? I found it always odd to call everything "cert" but then having a "certificatetpr" project.

File: searcher.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+func (s *Searcher) SearchCluster(clusterID string) (Cluster, error) {
+	var cluster Cluster
+	var err error
+
+	err = s.searchError(&cluster.APIServer, clusterID, apiCert, err)
+	err = s.searchError(&cluster.CalicoClient, clusterID, calicoCert, err)
+	err = s.searchError(&cluster.EtcdServer, clusterID, etcdCert, err)
+	err = s.searchError(&cluster.ServiceAccount, clusterID, serviceAccountCert, err)
+	err = s.searchError(&cluster.Worker, clusterID, workerCert, err)
```
Comment (xh3b4sd) [inline searcher.go]: IMO there should be the usual error handling or a list of assets we iterate over. Then we would only type the error handling ones. I do not like calling functions with errors to just make then return the error. That is unnecessary. We can do better.

File: searcher_test.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+				},
+			},
+			ExpectedTLS:      TLS{},
+			ExpectedErrMatch: IsInvalidSecret,
+		},
+	}
+
+	for i, tc := range testCases {
+		var tls TLS
+		err := fillTLSFromSecret(&tls, tc.Secret, tc.ClusterID, tc.Cert)
```
Comment (xh3b4sd) [inline searcher_test.go]: I personally do not like working with pointers that way. It is more straight forward to call a function to get some value back.

**Comment by kopiczko** [comment]
@xh3b4sd would `certs` be a better name for this project?

**Comment by xh3b4sd** [comment]
Not sure if it would be "better", but it would align better with the naming of the other projects.

**Comment by kopiczko** [comment]
`s/better/preffered/`. I'll rename the repo. It's not imported anywhere yet.

File: searcher_test.go (hunk 3 above)
Comment (kopiczko) [reply searcher_test.go]: I also favour returning values over passing pointers and mutating them. I just couldn't make better design. This is a private function - we can change it any time. Please feel free to provide a PR.

File: searcher.go (hunk 2 above)
Comment (kopiczko) [reply searcher.go]: I'll try with loop.

Comment (kopiczko) [reply searcher.go]: The loop is there.

**Pull request merged**